src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+    }
+
+    public Node getNode(int index) {
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            if (index < (int) (size / 2)) {
+                int currentIndex = 0;
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                return (Node<T>) currentNode;
+            } else  {
+                int currentIndex = size - 1;
+                Node currentNode = tail;
+                while (currentIndex != index) {
+                    currentNode = currentNode.prev;
+                    currentIndex--;
+                }
+                return (Node<T>) currentNode;
+            }
+
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            T previousNode = (T) getNode(index).value;
+            getNode(index).value = value;
+            return previousNode;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else if (size == 1) {
+            Node removedNode = head;
+            head = null;
+            size--;
+            return (T) removedNode.value;
+        } else if (index == 0) {
+            Node removedNode = head;
+            head.next.prev = null;
+            head = head.next;
+            size--;
+            if (removedNode == null) {
+                return null;
+            }
+            return (T) removedNode.value;
+        } else if (index == size - 1) {
+            Node removedNode = tail;
+            tail.prev.next = null;
+            tail = tail.prev;
+            size--;
+            return (T) removedNode.value;
+        } else {
+            int currentIndex = 0;
+            Node currentNode = head;
+            while (currentIndex != index) {
+                currentNode = currentNode.next;
+                currentIndex++;
+            }
+            Node removedNode = currentNode;
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+            size--;
+            return (T) removedNode.value;
+        }
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node currentNode = head;
```

Comment:
Common mistake: Don't forget that you can have `null` as a method parameter in the `remove(T object)` method. In this case, you should check if `object` is `null` before calling `equals` on it to avoid a `NullPointerException`.